src/tribler-core/tribler_core/modules/tunnel/community/triblertunnel_community.py

```diff
@@ -563,6 +578,95 @@ async def unload(self):
     def get_lookup_info_hash(self, info_hash):
         return hashlib.sha1(b'tribler anonymous download' + hexlify(info_hash).encode('utf-8')).digest()
 
+    @tc_lazy_wrapper_unsigned(HTTPRequestPayload)
+    async def on_http_request(self, source_address, payload, circuit_id):
+        if circuit_id not in self.exit_sockets:
+            self.logger.warning("Received unexpected http-request")
+            return
+        if len([cache for cache in self.request_cache._identifiers.values()
+                if isinstance(cache, HTTPRequestCache) and cache.circuit_id == circuit_id]) > 5:
+            self.logger.warning("Too many HTTP requests coming from circuit %s")
+            return
+
+        self.logger.debug("Got http-request from %s", source_address)
+        target_address = decode_address(payload.target)
+
+        try:
+            with async_timeout.timeout(3):
+                self.logger.debug("Opening TCP connection to %s", target_address)
+                reader, writer = await open_connection(*target_address)
+                writer.write(payload.request)
+                response = b''
+                while True:
+                    line = await reader.readline()
+                    response += line
+                    if not line.strip():
+                        # Read HTTP response body (1MB max)
+                        response += await reader.read(1024**2)
+                        break
+        except AsyncTimeoutError:
+            self.logger.warning('Tunnel HTTP request timed out')
+            return
+        finally:
+            writer.close()
+
+        # Note that, depending on the libtorrent version, bdecode does not always raise
+        # an exception, sometimes it returns None instead.
+        try:
+            response_decoded = lt.bdecode(response.split(b'\r\n\r\n')[1])
+        except (IndexError, RuntimeError):
+            response_decoded = None
+
+        if response_decoded is None:
+            self.logger.warning('Tunnel HTTP request not allowed')
+            return
+
+        num_cells = math.ceil(len(response) / MAX_HTTP_PACKET_SIZE)
```

Comment:
Why doesn't this crash like DeepCode's report suggests?

Reply:
I've seen the warning, but in Python 3 `Math.ceil` seems to produce an integer (also see the [docs](https://docs.python.org/3/library/math.html#math.ceil)).

Reply:
Accepted, checks out, weird though:

```python
import math
print(3/5, type(3/5))
> 0.6 <class 'float'>
print(math.ceil(3/5), type(math.ceil(3/5)))
> 1 <class 'int'>
```

According to the docs: "**If x is not a float**, delegates to x.__ceil__(), which should return an Integral value." The input to `math.ceil` is clearly a float though..